### PR TITLE
Token Distributor

### PR DESCRIPTION
**Pull request by cmontecoding**
Add contract to distribute early KWENTA vesting fees

### CODE REVIEW

File: contracts/TokenDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    /// can only be called once per week
+    function newDistribution() public {
+        ///@dev [epoch - 1] to get the start of last weeks epoch
+        if (epoch > 0) {
+            require(
+                block.timestamp >=
+                    (distributionEpochs[epoch - 1].epochStartTime + 1 weeks),
+                "TokenDistributor: Last week's epoch has not ended yet"
+            );
+        }
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:68]: Seems like epoch 0 can be created instantly from deployment, which would be wrong.

Epoch 0 should be a week from deployment.

File: contracts/TokenDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    /// @notice this function will fetch StakingRewardsV2 to see what their staked balance
+    /// was at the start of the epoch then calculate proportional fees and transfer to user
+    function claimDistribution(address to, uint epochNumber) public {
+        /// @dev if this is the first claim of a new epoch, call newDistribution to start a new epoch
+        if (
+            block.timestamp >=
+            (distributionEpochs[epoch - 1].epochStartTime + 604800)
+        ) {
+            newDistribution();
+        }
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:93]: Do you need this if statement if newDistribution already checks this?

File: contracts/TokenDistributor.sol [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            block.timestamp,
+            claimedFees,
+            kwenta.balanceOf(address(this))
+        );
+        //for a given block it might not be the final value of the block
+        distributionEpochs[epoch] = distribution;
+        //todo: timestamp is put in for block but that is technically the time
+        emit NewEpochCreated(block.timestamp, epoch);
+
+        epoch++;
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:81]: you can `emit NewEpochCreated(block.timestamp, epoch++);` it's the same thing

File: contracts/TokenDistributor.sol [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        require(
+            epochNumber < (epoch - 1),
+            "TokenDistributor: Epoch is not ready to claim"
+        );
+        //make sure its not the same block as the EpochStartNumber
+        require(
+            block.number !=
+                distributionEpochs[epochNumber].epochStartBlockNumber,
+            "Cannot claim in a new distribution block"
+        );
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:105]: Can you change this comment to summarize *why* this require is important?

File: contracts/TokenDistributor.sol [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        if (
+            block.timestamp >=
+            (distributionEpochs[epoch - 1].epochStartTime + 604800)
+        ) {
+            newDistribution();
+        }
+        //require the other cases
+        //add custom error (saves gas)
+        require(
+            epochNumber < (epoch - 1),
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:97]: This math don't check out

For example, if epoch 0 is posted and someone tries to claim for epoch 0:
```
epochNumber: 0
epoch: 1
(epoch - 1): 0

epochNumber < (epoch - 1) would return false
```

Has this claim distribution function been tested?

File: contracts/TokenDistributor.sol (hunk 2 above)
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:93]: i wasnt completely sure but i figured since newDistribution has a require - if its not ready for a new distribution - it would fail in every claim that isnt making a new distribution

File: contracts/TokenDistributor.sol (hunk 3 above)
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:81]: i emit before incrementing because the uint256 epoch actually leads ahead, so it always the next epoch to start or the one that will happen after the current one. but i've been thinking about changing this because it can be confusing and relates to the comments above and below.

File: contracts/TokenDistributor.sol (hunk 4 above)
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:105]: yeah it was meant to be a //todo: that i forgot to remove but that should be explained why anyway as well

File: contracts/TokenDistributor.sol (hunk 5 above)
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:97]: related to the above comment. Yes its been tested but it functions to the above logic. You can't claim for epoch 0 until epoch 1, but while you are in the middle of epoch 1 the uint256 epoch is actually == 2. So to clarify: epoch-1 is the current epoch you are going through and epoch-2 is the soonest epoch you can claim for. 

The numbering could likely be made less confusing, but it's mostly for the logic that needs data for fees recorded at the START of an epoch. Thinking of it now though I think I know how to fix it (or at least make it clearer).

File: contracts/TokenDistributor.sol (hunk 1 above)
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:68]: Yeah related to below. I have it backwards where it records from the start, but I think I can reverse it to make it better.

File: contracts/TokenDistributor.sol [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            return totalStaked;
+        }
+        uint256 proportionalFees = ((tokensPerEpoch[thisWeek] * userStaked) /
+            totalStaked);
+
+        return proportionalFees;
+    }
+
+    /// @notice function for calculating the start of a week with an offset
+    function startOfWeek(uint timestamp) public view returns (uint) {
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:195]: Should be internal if never planned to be called publicly

ie. utility functions don't need to be public

File: contracts/interfaces/ITokenDistributor.sol [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    error CannotClaimInNewEpochBlock();
+
+    /// @notice error when user tries to claim for same epoch twice
+    error CannotClaimTwice();
+
+    /// @notice error when user tries to claim for epoch with 0 staking
+    error NothingStakedThatEpoch();
+
+    /// @notice error when user tries to claim 0 fees
+    error CannotClaim0Fees();
```
Comment (JChiaramonte7) [inline contracts/interfaces/ITokenDistributor.sol:42]: Errors before functions see [S2](https://github.com/Kwenta/kwenta-solcurity)

File: contracts/TokenDistributor.sol [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    ) {
+        kwenta = IKwenta(_kwenta);
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+
+        /// @notice custom start day (startTime + param)
+        if (_offset > 6) {
+            revert OffsetTooBig();
+        }
+        offset = _offset * 86400;
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:58]: `1 days`

File: contracts/TokenDistributor.sol [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        uint256 proportionalFees = ((tokensPerEpoch[epochStart] * userStaked) /
+            totalStaked);
+
+        return proportionalFees;
+    }
+
+    /// @notice function for calculating the start of a week with an offset
+    function _startOfWeek(uint timestamp) internal view returns (uint) {
+        return (((timestamp - offset) / 1 weeks) * 1 weeks) + offset;
+    }
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:172]: @dev comment here would be good to explain why you subtract offset

**Review by JChiaramonte7 — CHANGES_REQUESTED** [review]

File: contracts/TokenDistributor.sol [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+        claimedEpochs[to][epochNumber] = true;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:139]: @tommyrharper is the 90% fee specified as 18 decimal int, basis points, or just plain integer as it is here?

**Comment by JChiaramonte7** [comment]
@tommyrharper @JaredBorders can you guys also review? code is pretty much close to completion

File: contracts/TokenDistributor.sol [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+        address _stakingRewardsV2,
+        address _rewardEscrowV2,
+        uint _offset
+    ) {
+        kwenta = IKwenta(_kwenta);
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+
+        /// @notice custom start day (startTime + param)
+        if (_offset > 6) {
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:49]: It is slightly confusing that offset in the constructor is in days but elsewhere offset is in seconds. I'd recommend renaming the constructor param

Comment (JaredBorders) [reply contracts/TokenDistributor.sol:49]: In fact, is it absolutely necessary to set the offset and many of these other related state variables in the constructor?

File: contracts/TokenDistributor.sol [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        offset = _offset * 1 days;
+        uint startOfThisWeek = _startOfWeek(block.timestamp);
+        startTime = startOfThisWeek;
+        lastCheckpoint = startOfThisWeek;
+    }
+
+    function checkpointToken() public override {
+        uint tokenBalance = kwenta.balanceOf(address(this));
+        uint toDistribute = tokenBalance - lastTokenBalance;
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:60]: Is there a test showing there can never be an underflow here? Can that happen theoretically?

File: contracts/TokenDistributor.sol [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        }
+
+        uint256 proportionalFees = calculateEpochFees(to, epochNumber);
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+        claimedEpochs[to][epochNumber] = true;
```
[the remark this answers is not in the document]
Comment (JaredBorders) [reply contracts/TokenDistributor.sol:128]: The update to `claimedEpochs` is what I am referencing specifically

File: contracts/TokenDistributor.sol [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+        uint previousCheckpoint = lastCheckpoint;
+        uint sinceLast = block.timestamp - previousCheckpoint;
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52
+        for (uint i = 0; i < 52; i++) {
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:71]: maybe make a constant like `weeksInYear` 
Bad practice having plain numbers floating around

File: contracts/TokenDistributor.sol [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+        if (proportionalFees == 0) {
+            revert CannotClaim0Fees();
+        }
+
+        lastTokenBalance -= proportionalFees;
+        claimedEpochs[to][epochNumber] = true;
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:131]: maybe make a constant like `weeksInYear` 
Bad practice having plain numbers floating around

File: contracts/TokenDistributor.sol [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+    /// @notice starting week of deployment
+    uint public startTime;
+
+    /// @notice array for tokens allocated to each epoch
+    uint[1000000000000000] public tokensPerEpoch;
+
+    /// @notice the week offset in seconds
+    uint public offset;
+
+    constructor(
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:38]: Add docs here

File: contracts/TokenDistributor.sol [hunk 17]
```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: MIT
```
Comment (JaredBorders) [inline contracts/TokenDistributor.sol:1]: Just realized we use `// SPDX-License-Identifier: GPL-3.0-or-later` in margin-manager. @JChiaramonte7 not sure if this matters or not

File: contracts/TokenDistributor.sol (hunk 11 above)
Comment (JaredBorders) [reply contracts/TokenDistributor.sol:49]: 6 should be a constant

**Comment by JaredBorders** [comment]
Also, are you generating code coverage for this locally? That should be shared in this repo somewhere

File: contracts/TokenDistributor.sol (hunk 17 above)
Comment (JChiaramonte7) [reply contracts/TokenDistributor.sol:1]: Yes good catch, we're moving to `// SPDX-License-Identifier: GPL-3.0-or-later` moving forward

File: contracts/TokenDistributor.sol (hunk 15 above)
Comment (JChiaramonte7) [reply contracts/TokenDistributor.sol:131]: I think we told him to do this (over his seconds approach), but yeah I do recommend using constants for anything you **reuse**. For example a constant called `uint YEAR = 52 weeks` if you use it a couple times.

In this case I think you're only using it once so it's up to you if you want to create a constant for it

File: contracts/TokenDistributor.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (JChiaramonte7) [reply contracts/TokenDistributor.sol:128]: @cmontecoding Checks-Effects-Interactions (CEI) pattern for context

https://docs.soliditylang.org/en/v0.6.11/security-considerations.html#re-entrancy

File: contracts/TokenDistributor.sol [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+    uint public lastTokenBalance;
+
+    /// @notice last checkpoint time
+    uint public lastCheckpoint;
+
+    /// @notice starting week of deployment
+    uint public startTime;
+
+    /// @notice array for tokens allocated to each epoch
+    uint[1000000000000000] public tokensPerEpoch;
```
[the remark this answers is not in the document]
Comment (cmontecoding) [reply contracts/TokenDistributor.sol:33]: I took that from Velo because thats what they use for their checkpointing system. I added a gas snapshot and the current deployment cost is 760K gas (not sure if thats a lot). I tried testing other ways to initialize but it just bricks checkpointToken because of the way Velo does it. Thoughts on this @JChiaramonte7?

[the remark this answers is not in the document]
Comment (tommyrharper) [reply contracts/TokenDistributor.sol:33]: I think you can use a mapping: `mapping(uint256 => uint256)`.
But curious to see if the above offers some gas advantage 🤔

File: test/foundry/unit/TokenDistributor.t.sol [hunk 19]
```diff
@@ -0,0 +1,1272 @@
...
+
+        uint proportionalFees3 = (amount * staking3) /
+            (staking1 + staking2 + staking3);
+        vm.expectEmit(true, true, true, true);
+        emit EpochClaim(address(user3), 1, proportionalFees3);
+        tokenDistributorOffset.claimEpoch(address(user3), 1);
+    }
+
+    /// @notice test fuzz time with a custom offset
+    function testFuzzTimeOffset(
```
Comment (tommyrharper) [inline test/foundry/unit/TokenDistributor.t.sol:787]: **issue:** This test is failing with the following args:
```
[3588613862 [3.588e9], 3202735154 [3.202e9], 3, 5225, 2155805 [2.155e6]]
```

File: contracts/TokenDistributor.sol [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+        /// this condition is so that the end of a week always
+        /// gets updated before its claimed (even if < 24 hrs)
+        if ((block.timestamp - _startOfWeek(lastCheckpoint)) > 1 weeks) {
+            checkpointToken();
+        }
+    }
+
+    /// @notice function for determining if the epoch being claimed
+    /// is the current epoch or has not happened yet
+    function _isEpochReady(uint epochNumber) internal {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:206]: **Issue:** This function can be restricted to type `view`.
```suggestion
    function _isEpochReady(uint epochNumber) internal view {
```

File: test/foundry/utils/TokenDistributorInternals.sol [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+    function startOfEpoch(uint epochNumber) public view returns (uint) {
+        return _startOfEpoch(epochNumber);
+    }
+
+    function checkpointWhenReady() public {
+        _checkpointWhenReady();
+    }
+
+    function isEpochReady(uint epochNumber) public {
```
Comment (tommyrharper) [inline test/foundry/utils/TokenDistributorInternals.sol:26]: **Issue:** This function can be restricted to type `external view`.
```suggestion
    function isEpochReady(uint epochNumber) external view {
```

File: contracts/TokenDistributor.sol [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52 weeks
+        for (uint i = 0; i < WEEKS_IN_YEAR; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:88]: **Thought:** Is it not correct that if `sinceLast == 0` then `block.timestamp == previousCheckpoint` by definition. Therefore only one of these conditions is needed?

File: contracts/TokenDistributor.sol [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+    uint internal lastTokenBalance;
+
+    /// @notice last checkpoint time
+    uint internal lastCheckpoint;
+
+    /// @notice starting week of deployment
+    uint internal startTime;
+
+    /// @inheritdoc ITokenDistributor
+    uint[1000000000000000] public tokensPerEpoch;
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:33]: **Thought/Issue:** Is it correct that currently `tokensPerEpoch` follows the following format:
```
startOfEpochTimestamp => numberOfTokensForGivenEpoch
```

I find it slightly unnerving that we use the timestamp as the key, rather than  what you would expect `epochNumber` to be.
It seems prone to errors of slightly incorrect timestamps than using an actual epoch number - i.e. epoch 1, epoch 2, epoch 3 etc.
Instead it is more like:
```
epochNumber 0 = startingTimestamp
epochNumber 1 = startingTimestamp + 604800
epochNumber 2 = startingTimestamp + 1209600
epochNumber 4 = startingTimestamp + 1814400
```
And so on.

I would prefer it if the epochNumbers being used were just "1, 2, 3, 4 etc.".

File: contracts/TokenDistributor.sol [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+                    /// for multiple weeks we take the remaining portion
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute *
+                            (block.timestamp - previousCheckpoint)) /
+                        sinceLast;
+                }
+                break;
+            } else {
+                /// @dev If passed weeks missed
+                if (sinceLast == 0 && nextWeek == previousCheckpoint) {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:103]: **Issue:** If `sinceLast == 0`, then surely it is not possible to be in this else block?
Because if that is the case, then the last checkpoint was in the very same second as now. And therefore `block.timestamp < nextWeek` cannot be true?

As if `block.timestamp == previousCheckpoint`, then `thisWeek` must be at the very smallest `block.timestamp - 604799`, and `nextWeek = block.timestamp + 604800`, therefore if `sinceLast = 0`, then `nextWeek` must be at least 1 greater than `block.timestamp`.

This means that `block.timestamp < nextWeek` will never evaluate as true if `sinceLast == 0` and hence this if block is unnecessary.

Definitely double check my logic, but pretty sure about this.

Comment (tommyrharper) [inline contracts/TokenDistributor.sol:103]: **Issue:** If `sinceLast == 0`, then surely it is not possible to be in this else block?
Because if that is the case, then the last checkpoint was in the very same second as now. And therefore `block.timestamp < nextWeek` cannot be true.

As if `block.timestamp == previousCheckpoint`, then `thisWeek` must be at the very smallest `block.timestamp - 604799`, and `nextWeek = block.timestamp + 604800`, therefore if `sinceLast = 0`, then `nextWeek` must be at least 1 greater than `block.timestamp`.

This means that `block.timestamp < nextWeek` will always evaluate as true if `sinceLast == 0` and hence this if block is unnecessary.

Definitely double check my logic, but pretty sure about this.

File: contracts/TokenDistributor.sol [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+
+        kwenta.approve(address(rewardEscrowV2), proportionalFees);
+        rewardEscrowV2.createEscrowEntry(to, proportionalFees, 52 weeks, 90);
+
+        emit EpochClaim(to, epochNumber, proportionalFees);
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function claimMany(address to, uint[] memory epochs) public {
+        for (uint i = 0; i < epochs.length; i++) {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:150]: **Suggestion:** This can be optimised in a few ways:
```solidity
uint256 length = epochs.length; // this is more efficient as length isn't accessed each iteration
for (uint i = 0; i < length;) {
  // do stuff
  unchecked { // unchecked is more efficient as it doesn't check for overflows
    ++i; // notice here ++i is more efficient than i++
  }
}
```

File: contracts/TokenDistributor.sol [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+        if (claimedEpochs[to][epochNumber] == true) {
+            revert CannotClaimTwice();
+        }
+        claimedEpochs[to][epochNumber] = true;
+
+        uint256 totalStaked = stakingRewardsV2.totalSupplyAtTime(
+            _startOfEpoch(epochNumber)
+        );
+        if (totalStaked == 0) {
+            revert NothingStakedThatEpoch();
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:131]: **Issue:** You also query `totalSupplyAtTime` in `calculateEpochFees` and `claimEpoch`.

This check is unnecessary as you have the following check in `calculateEpochFees`:
```solidity
        if (totalStaked == 0) {
            return 0;
        }
```

Hence you don't need the check or error `NothingStakedThatEpoch` at all, as `CannotClaim0Fees` will be triggered in that case.

This one is particularly important as `totalSupplyAtTime` could be very expensive as the number of checkpoints to binary search grows. So best not to call it twice if not needed.

File: contracts/TokenDistributor.sol [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+
+    /// @notice function for calculating the start of an epoch
+    function _startOfEpoch(uint epochNumber) internal view returns (uint) {
+        return (epochNumber * 1 weeks) + startTime;
+    }
+
+    /// @notice function for determining if a checkpoint is necessary
+    function _checkpointWhenReady() internal {
+        /// @dev if more than 24 hours from last checkpoint
+        if ((block.timestamp - lastCheckpoint > 1 days)) {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:193]: **Issue:** I don't believe there are any tests that check the edge case where `block.timestamp - lastCheckpoint == 1 days`, we should test this situation.

File: contracts/TokenDistributor.sol [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+            checkpointToken();
+        }
+        /// @dev if it is the first claim of the week then checkpoint.
+        /// this condition is so that the end of a week always
+        /// gets updated before its claimed (even if < 24 hrs)
+        if ((block.timestamp - _startOfWeek(lastCheckpoint)) > 1 weeks) {
```
Comment (tommyrharper) [inline contracts/TokenDistributor.sol:199]: **Issue:** I don't believe there are any tests that check the edge case where `block.timestamp - _startOfWeek(lastCheckpoint) == 1 weeks`, we should test this situation.

**Review by tommyrharper — CHANGES_REQUESTED** [review]
Overall looks good, but left a few comments on a couple of issues :)

File: contracts/TokenDistributor.sol (hunk 22 above)
Comment (tommyrharper) [reply contracts/TokenDistributor.sol:88]: Actually I think I am wrong here, mybad.

File: contracts/TokenDistributor.sol [hunk 29]
```diff
@@ -75,24 +75,31 @@ contract TokenDistributor is ITokenDistributor {
...
                 if (sinceLast == 0) {
                     /// @dev If no time change since last checkpoint just add new tokens
                     /// that may have been deposited (same block)
-                    tokensPerEpoch[thisWeek] += toDistribute;
+                    tokensPerEpoch[thisEpoch] += toDistribute;
                 } else {
                     /// @dev In the event that toDistribute contains tokens
                     /// for multiple weeks we take the remaining portion
-                    tokensPerEpoch[thisWeek] +=
+                    tokensPerEpoch[thisEpoch] +=
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol]: Instead of creating a whole new `nextEpoch` variable you need to keep track of (risking errors) I would just do the conversion right before you store the value for example
```solidity
uint epoch = (thisWeek - startTime) / 1 weeks;
tokensPerEpoch[epoch] += toDistribute;
```

also break it out into a function called `epochFromTimestamp(uint timestamp)`

File: contracts/TokenDistributor.sol [hunk 30]
```diff
@@ -0,0 +1,215 @@
...
+    mapping(address => mapping(uint => bool)) internal claimedEpochs;
+
+    /// @notice kwenta interface
+    IKwenta private kwenta;
+
+    /// @notice rewards staking contract
+    StakingRewardsV2 private stakingRewardsV2;
+
+    /// @notice escrow contract which holds (and may stake) reward tokens
+    RewardEscrowV2 private rewardEscrowV2;
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:21]: Can't really see a reason to make these private, internal is probably good enough in case you want to inherit this contract in the future

Comment (JChiaramonte7) [reply contracts/TokenDistributor.sol:21]: See [V10](https://github.com/Kwenta/kwenta-solcurity#variables)

File: contracts/TokenDistributor.sol [hunk 31]
```diff
@@ -0,0 +1,215 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.19;
+
+import {StakingRewardsV2} from "./StakingRewardsV2.sol";
+import {RewardEscrowV2} from "./RewardEscrowV2.sol";
+import {IKwenta} from "./interfaces/IKwenta.sol";
+import {ITokenDistributor} from "./interfaces/ITokenDistributor.sol";
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:7]: nit: Imports should be alphabetical

File: contracts/TokenDistributor.sol [hunk 32]
```diff
@@ -0,0 +1,215 @@
...
+    uint internal lastTokenBalance;
+
+    /// @notice last checkpoint time
+    uint internal lastCheckpoint;
+
+    /// @notice starting week of deployment
+    uint internal startTime;
+
+    /// @inheritdoc ITokenDistributor
+    mapping (uint => uint) public tokensPerEpoch;
```
Comment (JChiaramonte7) [inline contracts/TokenDistributor.sol:33]: Group your mappings together, also could make sense to group your variables by visibility. It took me a while to catch that this variable is the only one that's public.

File: test/foundry/unit/TokenDistributorGasCalculation.t.sol [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+/// and how much gas will it cost
+contract TokenDistributorGasCalculation is StakingSetup {
+
+    TokenDistributor public tokenDistributor;
+
+    function setUp() public override {
+        /// @dev starts after a week so the startTime is != 0
+        goForward(1 weeks + 1);
+        super.setUp();
+        switchToStakingV2();
```
Comment (JChiaramonte7) [inline test/foundry/unit/TokenDistributorGasCalculation.t.sol:23]: Do you need `switchToStakingV2()` for any of your tests? My understanding is that your function redirects inflation schedule to StakingV2 but that's only useful if you're minting new rewards.

File: test/foundry/unit/TokenDistributor.t.sol [hunk 34]
```diff
@@ -0,0 +1,1446 @@
...
+        assertEq(sum, 111);
+    }
+
+    /// @notice fuzz calculateEpochFees, especially when checkpoints are across weeks
+    function testFuzzCalculateEpochFees(uint256 amount) public {
+        /// @dev make sure its less than this contract
+        /// holds and greater than 10 so the result isn't
+        /// 0 after dividing
+        vm.assume(amount < 100_000 ether);
+        vm.assume(amount > 10);
```
Comment (JChiaramonte7) [inline test/foundry/unit/TokenDistributor.t.sol:304]: Does the contract have issues if the divisible amount of tokens is below 10? Meaning are those < 10 KWENTAs lost forever?

File: test/foundry/unit/TokenDistributor.t.sol [hunk 35]
```diff
@@ -0,0 +1,1446 @@
...
+    /// make sure theres no error dividing by 0
+    /// because thisEpoch should be 0
+    function testCheckpointTokenFirstWeek() public {
+        tokenDistributor.checkpointToken();
+        goForward(1 days);
+        tokenDistributor.checkpointToken();
+    }
+
+    /// @notice claimEpoch happy case
+    function testClaimEpoch() public {
```
Comment (JChiaramonte7) [inline test/foundry/unit/TokenDistributor.t.sol:83]: Missing a tests for > 1 person claiming and making sure each person gets expected amount

File: test/foundry/unit/TokenDistributor.t.sol [hunk 36]
```diff
@@ -0,0 +1,1446 @@
...
+    function testFuzzEpochsArent(uint8 epochNumber) public {
+        TokenDistributorInternals tDI = new TokenDistributorInternals(
+            address(kwenta),
+            address(stakingRewardsV2),
+            address(rewardEscrowV2),
+            0
+        );
+        /// @dev 15 epochs will already be claimable
+        vm.assume(epochNumber > 15);
+        goForward(15 weeks);
```
Comment (JChiaramonte7) [inline test/foundry/unit/TokenDistributor.t.sol:1371]: Why not fuzz from 0? Goal of fuzzing is to find any possible edge case, so it's better to not have more restriction than necessary.

**Review by JChiaramonte7 — COMMENTED** [review]
Put your tests into TokenDistributor folder
image

File: test/foundry/unit/TokenDistributorGasCalculation.t.sol (hunk 33 above)
Comment (cmontecoding) [reply test/foundry/unit/TokenDistributorGasCalculation.t.sol]: Doesn't all of TokenDistributor rely on StakingV2 because of the V2 checkpointing? Or you mean why did I choose to implement it in my tests by doing switchToStakingV2?

File: test/foundry/unit/TokenDistributor.t.sol (hunk 34 above)
Comment (cmontecoding) [reply test/foundry/unit/TokenDistributor.t.sol]: In situations where there is < 10 KWENTA and theres a large amount of staking and/or lots of stakers (compared to tokens), yes those KWENTA can't be claimed and are lost forever.

File: scripts/DeployTokenDistributor.sol
Comment (JaredBorders) [inline scripts/DeployTokenDistributor.sol:1]: Will there not be a testnet deployment?

**Review by JaredBorders — APPROVED** [review]
Left a few questions but LGTM otherwise
